contracts/governance/Staking/modules/StakingAdminModule.sol

```diff
@@ -0,0 +1,158 @@
+pragma solidity ^0.5.17;
+pragma experimental ABIEncoderV2;
+
+import "../../../openzeppelin/Address.sol";
+import "./shared/StakingShared.sol";
+import "../../../proxy/modules/interfaces/IFunctionsList.sol";
+
+/**
+ * @title Staking Admin Module.
+ * @notice Implements administrative functionality pause, freeze and setting addresses and parameters
+ * related to staking
+ * */
+contract StakingAdminModule is IFunctionsList, StakingShared {
+    using Address for address payable;
+
+    event AdminAdded(address admin);
+
+    event AdminRemoved(address admin);
+
+    /// @param pauser address to grant power to pause the contract
+    /// @param added true - added, false - removed
+    event PauserAddedOrRemoved(address indexed pauser, bool indexed added);
+
+    /// @notice An event emitted when a staking is paused or unpaused
+    /// @param setPaused true - pause, false - unpause
+    event StakingPaused(bool indexed setPaused);
+
+    /// @notice An event emitted when a staking is frozen or unfrozen
+    /// @param setFrozen true - freeze, false - unfreeze
+    event StakingFrozen(bool indexed setFrozen);
+
+    /**
+     * @notice Add account to Admins ACL.
+     * @param _admin The addresses of the account to grant permissions.
+     * */
+    function addAdmin(address _admin) external onlyOwner whenNotFrozen {
+        admins[_admin] = true;
+        emit AdminAdded(_admin);
+    }
+
+    /**
+     * @notice Remove account from Admins ACL.
+     * @param _admin The addresses of the account to revoke permissions.
+     * */
+    function removeAdmin(address _admin) external onlyOwner whenNotFrozen {
+        admins[_admin] = false;
+        emit AdminRemoved(_admin);
+    }
+
+    /**
+     * @notice Add account to pausers ACL.
+     * @param _pauser The address to grant pauser permissions.
+     * */
+    function addPauser(address _pauser) external onlyOwner whenNotFrozen {
+        pausers[_pauser] = true;
+        emit PauserAddedOrRemoved(_pauser, true);
+    }
+
+    /**
+     * @notice Remove account from pausers ACL.
+     * @param _pauser The address to grant pauser permissions.
+     * */
+    function removePauser(address _pauser) external onlyOwner whenNotFrozen {
+        delete pausers[_pauser];
+        emit PauserAddedOrRemoved(_pauser, false);
+    }
+
+    /**
+     * @notice Pause/unpause contract
+     * @param _pause true when pausing, false when unpausing
+     * */
+    function pauseUnpause(bool _pause) public onlyPauserOrOwner whenNotFrozen {
+        paused = _pause;
+        emit StakingPaused(_pause);
+    }
+
+    /**
+     * @notice Freeze contract - disable all functions
+     * @param _freeze true when freezing, false when unfreezing
+     * @dev When freezing, pause is always applied too. When unfreezing, the contract is left in paused stated.
+     * */
+    //TODO: separate freeze from unfreeze
+    function freezeUnfreeze(bool _freeze) external onlyPauserOrOwner {
```

Comment:
this was marked as resolved. would be good to have the reasoning here.
is it because with a SIP the contract could be unfrozen and the rogue admin removed?

Reply:
yes